single-kernel/lin_reg_coeff.cpp

```diff
@@ -0,0 +1,196 @@
+    input1.resize(args.problem_size);
+    input2.resize(args.problem_size);
+    output.resize(args.problem_size, 0);
+
+    input1ver.resize(args.problem_size);
+    input2ver.resize(args.problem_size);
+
+    for (size_t i = 0; i < args.problem_size; i++) {
+       input1ver[i] = input1[i] = 1.0;
+       input2ver[i] = input2[i] = 2.0;
+    }
+  }
+
+  void vec_product(s::buffer<T, 1> &input1_buf, s::buffer<T, 1> &input2_buf, s::buffer<T, 1> &output_buf) {
+    args.device_queue.submit(
+        [&](cl::sycl::handler& cgh) {
+      auto in1 = input1_buf.template get_access<s::access::mode::read>(cgh);
+      auto in2 = input2_buf.template get_access<s::access::mode::read>(cgh);
+ 
+       // Use discard_write here, otherwise the content of the host buffer must first be copied to device
+      auto intermediate_product = output_buf.template get_access<s::access::mode::discard_write>(cgh);
+
+      cl::sycl::nd_range<1> ndrange (args.problem_size, args.local_size);
+
+      cgh.parallel_for<class VecProductKernel<T>>(ndrange,
+        [=](cl::sycl::nd_item<1> item) 
+        {
+          size_t gid= item.get_global_linear_id();
+          intermediate_product[gid] = in1[gid] * in2[gid];
+        });
+    });
+  }
+
+T reduce(s::buffer<T, 1> &input_buf) {
+  auto array_size = args.problem_size;
+  auto wgroup_size = args.local_size;
+  // Not yet tested with more than 2
+  auto elements_per_thread = 2;
+
+  while (array_size!= 1) {
+    auto n_wgroups = (array_size + wgroup_size*elements_per_thread - 1)/(wgroup_size*elements_per_thread); // two threads per work item
+
+    args.device_queue.submit(
+      [&](cl::sycl::handler& cgh) {
+
+        auto global_mem = input_buf.template get_access<s::access::mode::read_write>(cgh);
+    
+        // local memory for reduction
+        auto local_mem = s::accessor <T, 1, s::access::mode::read_write, s::access::target::local> {s::range<1>(wgroup_size), cgh};
+        cl::sycl::nd_range<1> ndrange (n_wgroups*wgroup_size, wgroup_size);
+  
+        cgh.parallel_for<class VecReduceKernel<T>>(ndrange,
+        [=](cl::sycl::nd_item<1> item) 
+          {
+            size_t gid= item.get_global_linear_id();
+            size_t lid = item.get_local_linear_id();
+
+            // initialize local memory to 0
+            local_mem[lid] = 0; 
+
+            if ((elements_per_thread * gid) < array_size) {
+                local_mem[lid] = global_mem[elements_per_thread*gid] + global_mem[elements_per_thread*gid + 1];
+            }
+
+            item.barrier(s::access::fence_space::local_space);
+
+            for (size_t stride = 1; stride < wgroup_size; stride *= elements_per_thread) {
+              auto local_mem_index = elements_per_thread * stride * lid;
+              if (local_mem_index < wgroup_size) {
+                  local_mem[local_mem_index] = local_mem[local_mem_index] + local_mem[local_mem_index + stride];
+              }
+
+              item.barrier(s::access::fence_space::local_space);
```

Comment:
I think the reduction from the reduction benchmark may be more efficient because of memory bank conflicts and because it saves one barrier. We don't need to have the most efficient reduction of all time, but maybe we can agree on one way to formulate reductions such that the benchmarks reuse the same patterns?

Reply:
I was also thinking of this. Maybe we can move some commonly used functions to common file and just call them when needed. No need to replicate code.

Reply:
Yes, I think that's a good idea.

Reply:
Good then, we can discuss this in our next call and do it!